config.org
#+title:     Doom Emacs Literate Config
#+author:    Jay L. Colbert
#+email:     jaylcolbert@gmail.com
#+property: header-args :mkdirp yes

* Table of Contents :toc_3:
- [[#introduction][Introduction]]
  - [[#why-emacs][Why Emacs?]]
  - [[#using-this-config][Using this config]]
  - [[#current-issues][Current issues]]
  - [[#possible-future-features][Possible future features]]
- [[#rudimentary-configuration][Rudimentary Configuration]]
  - [[#personal-information][Personal information]]
  - [[#better-defaults][Better defaults]]
    - [[#simple-settings][Simple settings]]
    - [[#frame-sizing][Frame sizing]]
- [[#doom-configuration][Doom configuration]]
  - [[#modules][Modules]]
    - [[#completion][Completion]]
    - [[#ui][UI]]
    - [[#editor][Editor]]
    - [[#tools][Tools]]
    - [[#languages][Languages]]
  - [[#visual-settings][Visual settings]]
    - [[#font-face][Font face]]
    - [[#theme][Theme]]
    - [[#line-numbers][Line numbers]]
- [[#packages][Packages]]
  - [[#doct][Doct]]
  - [[#hydra-posframe][Hydra-posframe]]
  - [[#info-colors][Info-colors]]
  - [[#major-mode-hydra][Major-mode-hydra]]
  - [[#org-ol-tree][Org-ol-tree]]
  - [[#org-remark][Org-remark]]
  - [[#org-roam-desktop][Org-roam-desktop]]
  - [[#org-transclusion][Org-transclusion]]
  - [[#rotate][Rotate]]
  - [[#zetteldesk][Zetteldesk]]
- [[#documents][Documents]]
  - [[#configel][=config.el=]]
  - [[#packagesel][=packages.el=]]
  - [[#autoloads][Autoloads]]
    - [[#hydrasel][=hydras.el=]]
    - [[#iconsel][=icons.el=]]
    - [[#org-captureel][=org-capture.el=]]
    - [[#org-roamel][=org-roam.el=]]
    - [[#posframeel][=posframe.el=]]
    - [[#zetteldesk-remarkel][=zetteldesk-remark.el=]]

* Introduction
** Why Emacs?
** Using this config
If you want your setup exactly like mine, here are the necessary prerequisites.
All of these commands are run on the command line.

*Have a Mac*

   I'm on the M1 chip now.

*Install Homebrew*

#+begin_src sh :tangle no
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
#+end_src

*Install Emacs Plus*

   I am on Emacs 28, built with native compilation.
   You don't have to do native comp, but it's supposed to be faster I think.
   I also added a different icon, but that's not necessary.
#+begin_src sh :tangle no
brew tap d12frosted/emacs-plus
brew install emacs-plus@28 --with-native-comp --with-nobu417-big-sur-icon
ln -s /opt/homebrew/Cellar/emacs-plus@27/27.2/Emacs.app /Applications
#+end_src

*Install Doom*

   First we need some dependencies:
#+begin_src sh :tangle no
# required dependencies
brew install git ripgrep
# optional dependencies
brew install coreutils fd
# Installs clang
xcode-select --install
#+end_src

  Now we can install Doom.
  This will take forever because of the native comp.
#+begin_src sh :tangle no
git clone --depth 1 https://github.com/hlissner/doom-emacs ~/.config/emacs
~/.config/emacs/bin/doom install
#+end_src

  I put my configs in =~/.config= as much as possible.
  The documentation puts it in a =~/.emacs.d= folder.

  Then I add it to my ~PATH~.
  I use the fish shell, so my way of doing this will be different.
  Change as necessary for your shell.
#+begin_src sh :tangle no
fish_add_path $HOME/.config/emacs/bin
#+end_src

  Now we can use the ~doom~ utility.
  I recommend running ~doom sync~ first.

  My standard workflow after installing/removing packages or changing =init.el= is:
#+begin_src sh :tangle no
doom sync -up
doom purge -g
doom doctor
#+end_src

For the rest of my setup, please see my dotfiles.

** Current issues
** Possible future features
* Rudimentary Configuration
** Personal information
Enter in the info we want when things auto-fill names/emails.
I think this also involves GPG, but I'm not sure.
#+begin_src elisp :tangle no :noweb-ref personal-info
(setq user-full-name    "Jay L. Colbert"
      user-mail-address "jaylcolbert@gmail.com")
#+end_src

** Better defaults
*** Simple settings
I have ADHD, so I need to see the time and battery in the program.
Otherwise, I won't remember to look up at the time in my Mac menu bar.
#+begin_src elisp :tangle no :noweb-ref simple-settings
(display-time-mode 1)
(display-battery-mode 1)
#+end_src

I don't want to see =.DS_Store= files when I'm searching.
#+begin_src elisp :tangle no :noweb-ref simple-settings
(add-to-list 'completion-ignored-extensions ".DS_Store")
#+end_src

*** Frame sizing
I use =emacs-plus=.
It's great, but there's one things that =emacs-mac= did that I miss:
centering the initial frame and sizing it in a way I really liked.
Changing my settings to replicate that is takes a few steps.
According to the =initial-frame-alist= documentation:
#+begin_quote
You can specify geometry-related options for just the initial
frame by setting this variable in your init file; however, they
won't take effect until Emacs reads your init file, which happens
after creating the initial frame.  If you want the initial frame
to have the proper geometry as soon as it appears, you need to
use this three-step process:
+ Specify X resources to give the geometry you want.
+ Set default-frame-alist to override these options so that they
  don't affect subsequent frames.
+ Set initial-frame-alist in a way that matches the X resources,
  to override what you put in default-frame-alist.
#+end_quote

So let's do that.

Newer Macs don't use X to manage windows.
They use Cocoa (which has NSWindow).
Luckily, specifying this geometry is the same process.
It's been a little fussy, but it's good enough for me right now.
We set the width and height in characters and the position of the top left corner in pixels.
(So you need to figure out the pixel position.
I have a program that will center windows on my screen.
I did that, and then I went to take a screenshot but only for a specific section.
It gives you a little crosshair with the pixel position.
I put the crosshair there and used those pixel coordinates.)
#+begin_src sh :tangle no
defaults write org.gnu.Emacs Width 80
defaults write org.gnu.Emacs Height 40
defaults write org.gnu.Emacs Top 181
defaults write org.gnu.Emacs Left 422
#+end_src

Next, we can set =default-frame-alist= to override our geometry settings.
I don't care about the position of default frames.
I rarely use more than one frame, so the geometry defaults and the initial frame settings usually do the trick for me.
I also don't mind the size of the default frame when I do create additional ones.

Finally, it tells us to set =initial-frame-alist= to match our geometry settings.
The width and height parameters are set the same, but I think top and left are percentage/fractional.
These settings replicate my defaults above:
#+begin_src elisp :tangle no :noweb-ref frame-sizing
(setq initial-frame-alist
      '((top . 0.5) (left . 0.5) (width . 80) (height . 40)))
#+end_src

* Doom configuration
** Modules
This is where I configure the modules I've activated in =init.el=.
Other packages are configured elsewhere.
*** Completion
**** Company
I hate when the completion suggestions pop up super fast.
This gives a bit of a delay.
#+begin_src elisp :tangle no :noweb-ref company
(after! company
  (setq company-idle-delay 0.5))
#+end_src

*** UI
**** Deft
I have Deft enabled, but I need to do my config for it.
I don't really use it that much, but it's something that I know is already installed or something by another package, so I like to at least have the Doom configuration turned on.

**** Doom
These are the Doom settings meant to be configured outside of module configuration.
The first two just give the buffer a nicer looking name when there's nothing in it.
The third is from Tecosaur.
It puts some cool icons that indicate if something is an Org Roam file or a project file.
It also lets you know if you've saved your changes via some neat circles.
Finally, we have another delay setting.
This time, it speeds up the Which Key menu.
This variable says to set it before Which Key loads, which is why I'm not configuring it as its own module.
#+begin_src elisp :tangle no :noweb-ref doom
(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom"
      frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ?  buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p)  " ◉ %s" "  ●  %s") project-name)))))
      which-key-idle-delay 0.5
      +org-roam-auto-backlinks-buffer t)
#+end_src

**** Modeline
Another Tecosaur creation.
This improves the PDF information in the modeline and cleans up the buffer name.
#+begin_src elisp :tangle no :noweb-ref modeline
(after! doom-modeline
  (doom-modeline-def-segment buffer-name
    "Display the current buffer's name, without any other information."
    (concat
     (doom-modeline-spc)
     (doom-modeline--buffer-name)))

  (doom-modeline-def-segment pdf-icon
    "PDF icon from all-the-icons."
    (concat
     (doom-modeline-spc)
     (doom-modeline-icon 'octicon "file-pdf" nil nil
                         :face (if (doom-modeline--active)
                                   'all-the-icons-red
                                 'mode-line-inactive)
                         :v-adjust 0.02)))

  (defun doom-modeline-update-pdf-pages ()
    "Update PDF pages."
    (setq doom-modeline--pdf-pages
          (let ((current-page-str (number-to-string (eval `(pdf-view-current-page))))
                (total-page-str (number-to-string (pdf-cache-number-of-pages))))
            (concat
             (propertize
              (concat (make-string (- (length total-page-str) (length current-page-str)) ? )
                      " P" current-page-str)
              'face 'mode-line)
             (propertize (concat "/" total-page-str) 'face 'doom-modeline-buffer-minor-mode)))))

  (doom-modeline-def-segment pdf-pages
    "Display PDF pages."
    (if (doom-modeline--active) doom-modeline--pdf-pages
      (propertize doom-modeline--pdf-pages 'face 'mode-line-inactive)))

  (doom-modeline-def-modeline 'pdf
    '(bar window-number pdf-pages pdf-icon buffer-name)
    '(misc-info matches major-mode process vcs)))
#+end_src

*** Editor
**** Evil
Ensures a new window is always created to the right or below.
It also keeps the cursor from moving back when entering normal mode.
I do this to make adding links and citations easier for me.

When creating a new window, it will now prompt for the buffer or file you want to open.

And then, we add some keybindings to use arrows to navigate windows.
We can also rotate layouts.
#+begin_src elisp :tangle no :noweb-ref evil
(after! evil
  (setq evil-vsplit-window-right t
        evil-split-window-below t
        evil-move-cursor-back nil
        evil-kill-on-visual-paste nil)
  (defadvice! prompt-for-buffer (&rest _)
    :after '(evil-window-split evil-window-vsplit)
    (consult-buffer))
  (map! :map evil-window-map
        "SPC" #'rotate-layout
        ;; Navigation
        "<left>"     #'evil-window-left
        "<down>"     #'evil-window-down
        "<up>"       #'evil-window-up
        "<right>"    #'evil-window-right
        ;; Swapping windows
        "C-<left>"       #'+evil/window-move-left
        "C-<down>"       #'+evil/window-move-down
        "C-<up>"         #'+evil/window-move-up
        "C-<right>"      #'+evil/window-move-right))
#+end_src

*** Tools
**** Biblio
Because this is the configuration for my personal laptop, not my work laptop (I'm a librarian, but I also have faculty status), I don't do as much research on here as on there.
However, I am working on a second degree, so being able to write and cite is crucial.
I also have a Zettelkasten, and I like being able to cite with that.

Because I use Vertico, the bibliographic package to complement org-cite is citar.
All of these settings are from the Citar page.
#+begin_src elisp :tangle no :noweb-ref citar
(after! citar
  ;; use consult-completing-read for enhanced interface
  (advice-add #'completing-read-multiple :override #'consult-completing-read-multiple)
  (setq citar-bibliography      '("~/Documents/org/digital-garden/org/biblio.bib")
         citar-library-paths  '("~/Documents/Zotero")
         citar-symbols
         `((file ,(all-the-icons-faicon "file-o" :face 'all-the-icons-green :v-adjust -0.1) . " ")
           (note ,(all-the-icons-material "speaker_notes" :face 'all-the-icons-blue :v-adjust -0.3) . " ")
           (link ,(all-the-icons-octicon "link" :face 'all-the-icons-orange :v-adjust 0.01) . " "))
         citar-symbol-separator "  "))
#+end_src

*** Languages
**** Clojure
I mainly use Clojure for generative art.
My setup isn't that complicated because of it.

To be written (I haven't made art in a few months)...

**** Org
According to the Doom documentation, the Org directory needs to be set before org-mode loads.
Because I have that whole frame title thing that involves org-roam, I also load that immediately.
#+begin_src elisp :tangle no :noweb-ref org-dirs
(setq org-directory       "~/Documents/org/"
      org-roam-directory  (file-truename "~/Documents/org/digital-garden/org"))
#+end_src

Configuring org and the module flags I enabled is a little complicated.
I nest the configurations.
I think I'm allowed to do that.
#+begin_src elisp :tangle no :noweb-ref org :noweb yes
(after! org
  <<org-visuals>>
  <<org-capture>>
  <<org-journal>>
  <<org-roam>>)
#+end_src

***** Visuals
When I'm reading, I don't like to see the emphasis markers.
I also like having a fancy little triangle for the ellipsis in headings and other collapsible items.
#+begin_src elisp :tangle no :noweb-ref org-visuals
(setq org-hide-emphasis-markers t
      org-ellipsis " ▾ "
      org-startup-with-inline-images t
      org-image-actual-width '(300))
#+end_src

And then configuring this package means the emphasis markers will show back up when my cursor is inside that word/phrase.
I turned it off for links because it made navigating around my org-roam files really annoying.
#+begin_src elisp :tangle no :noweb-ref org-appear
(use-package! org-appear
  :hook (org-mode . org-appear-mode)
  :config
  (setq org-appear-autoemphasis t
        org-appear-autosubmarkers t
        org-appear-autolinks nil)
  ;; for proper first-time setup, `org-appear--set-elements'
  ;; needs to be run after other hooks have acted.
  (run-at-time nil nil #'org-appear--set-elements))
#+end_src

***** Capture
Alright, this is a big one.

Basically all of this is from tecosaur's config.
Because it's so complicated, I'm also breaking it up into pieces.
#+begin_src elisp :tangle no :noweb-ref org-capture :noweb yes
(after! org-capture
  <<prettify-capture>>
  <<org-capture-templates>>)
#+end_src

I autoload this function.
It's part of the settings that make the capture template popup look better.
#+begin_src elisp :tangle no :noweb-ref org-capture-select-template-prettier
(defun org-capture-select-template-prettier (&optional keys)
  "Select a capture template, in a prettier way than default
Lisp programs can force the template by setting KEYS to a string."
  (let ((org-capture-templates
         (or (org-contextualize-keys
              (org-capture-upgrade-templates org-capture-templates)
              org-capture-templates-contexts)
             '(("t" "Task" entry (file+headline "" "Tasks")
                "* TODO %?\n  %u\n  %a")))))
    (if keys
        (or (assoc keys org-capture-templates)
            (error "No capture template referred to by \"%s\" keys" keys))
      (org-mks org-capture-templates
               "Select a capture template\n━━━━━━━━━━━━━━━━━━━━━━━━━"
               "Template key: "
               `(("q" ,(concat (all-the-icons-octicon "stop" :face 'all-the-icons-red :v-adjust 0.01) "\tAbort")))))))
#+end_src

And same thing here.
#+begin_src elisp :tangle no :noweb-ref org-mks-pretty
(defun org-mks-pretty (table title &optional prompt specials)
  "Select a member of an alist with multiple keys. Prettified.

TABLE is the alist which should contain entries where the car is a string.
There should be two types of entries.

1. prefix descriptions like (\"a\" \"Description\")
   This indicates that `a' is a prefix key for multi-letter selection, and
   that there are entries following with keys like \"ab\", \"ax\"…

2. Select-able members must have more than two elements, with the first
   being the string of keys that lead to selecting it, and the second a

config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-

;; Place your private configuration here! Remember, you do not need to run 'doom
;; sync' after modifying this file!


;; Some functionality uses this to identify you, e.g. GPG configuration, email
;; clients, file templates and snippets. It is optional.
(setq user-full-name    "Jay L. Colbert"
      user-mail-address "jaylcolbert@gmail.com")

;; Doom exposes five (optional) variables for controlling fonts in Doom:
;;
;; - `doom-font' -- the primary font to use
;; - `doom-variable-pitch-font' -- a non-monospace font (where applicable)
;; - `doom-big-font' -- used for `doom-big-font-mode'; use this for
;;   presentations or streaming.
;; - `doom-unicode-font' -- for unicode glyphs
;; - `doom-serif-font' -- for the `fixed-pitch-serif' face
;;
;; See 'C-h v doom-font' for documentation and more examples of what they
;; accept. For example:
;;
(setq doom-font (font-spec :family "Fira Code" :size 12 :weight 'semi-light)
      doom-variable-pitch-font (font-spec :family "Fira Sans" :size 13))
;;
;; If you or Emacs can't find your font, use 'M-x describe-font' to look them
;; up, `M-x eval-region' to execute elisp code, and 'M-x doom/reload-font' to
;; refresh your font settings. If Emacs still can't find your font, it likely
;; wasn't installed correctly. Font issues are rarely Doom issues!

;; There are two ways to load a theme. Both assume the theme is installed and
;; available. You can either set `doom-theme' or manually load a theme with the
;; `load-theme' function. This is the default:
;;(setq doom-theme 'doom-one)
;;(setq doom-theme 'doom-earl-grey)
;;(setq doom-theme 'doom-flatwhite)
;;(setq doom-theme 'doom-nord)
;;(setq doom-theme 'doom-opera)
;;(setq doom-theme 'doom-opera-light)
;;(setq doom-theme 'doom-wilmersdorf)
;;(setq doom-theme 'doom-zenburn)
(defun jc/apply-theme (appearance)
  "Load theme, taking current system APPEARANCE into consideration."
  (mapc #'disable-theme custom-enabled-themes)
  (pcase appearance
    ('light (load-theme 'doom-earl-grey t))
    ('dark (load-theme 'doom-zenburn t))))

(add-hook 'ns-system-appearance-change-functions #'jc/apply-theme)

;; This determines the style of line numbers in effect. If set to `nil', line
;; numbers are disabled. For relative line numbers, set this to `relative'.
(setq display-line-numbers-type 'relative)

;; If you use `org' and don't want your org files in the default location below,
;; change `org-directory'. It must be set before org loads!
(setq org-directory       "~/Documents/org/"
      org-roam-directory  (file-truename "~/Documents/org/digital-garden/org"))


;; Whenever you reconfigure a package, make sure to wrap your config in an
;; `after!' block, otherwise Doom's defaults may override your settings. E.g.
;;
;;   (after! PACKAGE
;;     (setq x y))
;;
;; The exceptions to this rule:
;;
;;   - Setting file/directory variables (like `org-directory')
;;   - Setting variables which explicitly tell you to set them before their
;;     package is loaded (see 'C-h v VARIABLE' to look up their documentation).
;;   - Setting doom variables (which start with 'doom-' or '+').
;;
;; Here are some additional functions/macros that will help you configure Doom.
;;
;; - `load!' for loading external *.el files relative to this one
;; - `use-package!' for configuring packages
;; - `after!' for running code after a package has loaded
;; - `add-load-path!' for adding directories to the `load-path', relative to
;;   this file. Emacs searches the `load-path' when you load packages with
;;   `require' or `use-package'.
;; - `map!' for binding new keys
;;
;; To get information about any of these functions/macros, move the cursor over
;; the highlighted symbol at press 'K' (non-evil users must press 'C-c c k').
;; This will open documentation for it, including demos of how they are used.
;; Alternatively, use `C-h o' to look up a symbol (functions, variables, faces,
;; etc).
;;
;; You can also try 'gd' (or 'C-c c d') to jump to their definition and see how
;; they are implemented.

(display-time-mode 1)
(display-battery-mode 1)
(add-to-list 'completion-ignored-extensions ".DS_Store")

(setq initial-frame-alist
      '((top . 0.5) (left . 0.5) (width . 80) (height . 40)))

(setq doom-fallback-buffer-name "► Doom"
      +doom-dashboard-name "► Doom"
      frame-title-format
      '(""
        (:eval
         (if (s-contains-p org-roam-directory (or buffer-file-name ""))
             (replace-regexp-in-string
              ".*/[0-9]*-?" "☰ "
              (subst-char-in-string ?_ ?  buffer-file-name))
           "%b"))
        (:eval
         (let ((project-name (projectile-project-name)))
           (unless (string= "-" project-name)
             (format (if (buffer-modified-p)  " ◉ %s" "  ●  %s") project-name)))))
      which-key-idle-delay 0.5
      +org-roam-auto-backlinks-buffer t)

(after! citar
  ;; use consult-completing-read for enhanced interface
  (advice-add #'completing-read-multiple :override #'consult-completing-read-multiple)
  (setq citar-bibliography      '("~/Documents/org/digital-garden/org/biblio.bib")
         citar-library-paths  '("~/Documents/Zotero")
         citar-symbols
         `((file ,(all-the-icons-faicon "file-o" :face 'all-the-icons-green :v-adjust -0.1) . " ")
           (note ,(all-the-icons-material "speaker_notes" :face 'all-the-icons-blue :v-adjust -0.3) . " ")
           (link ,(all-the-icons-octicon "link" :face 'all-the-icons-orange :v-adjust 0.01) . " "))
         citar-symbol-separator "  "))

(after! company
  (setq company-idle-delay 0.5))

(use-package! doct
  :commands doct)

(after! doom-modeline
  (doom-modeline-def-segment buffer-name
    "Display the current buffer's name, without any other information."
    (concat
     (doom-modeline-spc)
     (doom-modeline--buffer-name)))

  (doom-modeline-def-segment pdf-icon
    "PDF icon from all-the-icons."
    (concat
     (doom-modeline-spc)
     (doom-modeline-icon 'octicon "file-pdf" nil nil
                         :face (if (doom-modeline--active)
                                   'all-the-icons-red
                                 'mode-line-inactive)
                         :v-adjust 0.02)))

  (defun doom-modeline-update-pdf-pages ()
    "Update PDF pages."
    (setq doom-modeline--pdf-pages
          (let ((current-page-str (number-to-string (eval `(pdf-view-current-page))))
                (total-page-str (number-to-string (pdf-cache-number-of-pages))))
            (concat
             (propertize
              (concat (make-string (- (length total-page-str) (length current-page-str)) ? )
                      " P" current-page-str)
              'face 'mode-line)
             (propertize (concat "/" total-page-str) 'face 'doom-modeline-buffer-minor-mode)))))

  (doom-modeline-def-segment pdf-pages
    "Display PDF pages."
    (if (doom-modeline--active) doom-modeline--pdf-pages
      (propertize doom-modeline--pdf-pages 'face 'mode-line-inactive)))

  (doom-modeline-def-modeline 'pdf
    '(bar window-number pdf-pages pdf-icon buffer-name)
    '(misc-info matches major-mode process vcs)))

(after! evil
  (setq evil-vsplit-window-right t
        evil-split-window-below t
        evil-move-cursor-back nil
        evil-kill-on-visual-paste nil)
  (defadvice! prompt-for-buffer (&rest _)
    :after '(evil-window-split evil-window-vsplit)
    (consult-buffer))
  (map! :map evil-window-map
        "SPC" #'rotate-layout
        ;; Navigation
        "<left>"     #'evil-window-left
        "<down>"     #'evil-window-down
        "<up>"       #'evil-window-up
        "<right>"    #'evil-window-right
        ;; Swapping windows
        "C-<left>"       #'+evil/window-move-left
        "C-<down>"       #'+evil/window-move-down
        "C-<up>"         #'+evil/window-move-up
        "C-<right>"      #'+evil/window-move-right))

(use-package! hydra-posframe
  :hook (after-init . hydra-posframe-mode)
  :config
  (setq hydra-posframe-parameters '((alpha 100 100)
                                    (left-fringe . 10)
                                    (right-fringe . 10))
        hydra-posframe-poshandler 'jc-frame-bottom-poshandler))

(use-package! info-colors
  :commands info-colors-fontify-node
  :init
  (add-hook 'Info-selection-hook #'info-colors-fontify-node))

(use-package major-mode-hydra
  :bind
  ("C-SPC" . major-mode-hydra)
  :config
  (setq major-mode-hydra-invisible-quit-key "q"
        major-mode-hydra-title-generator #'jc-major-mode-hydra-title-generator))

(major-mode-hydra-define emacs-lisp-mode nil
  ("Eval"
   (("b" eval-buffer "buffer")
    ("d" eval-defun "defun")
    ("e" eval-last-sexp "sexp")
    ("r" eval-region "region"))
   "REPL"
   (("I" ielm "ielm"))
   "Test"
   (("t" ert "prompt")
    ("T" (ert t) "all")
    ("F" (ert :failed) "failed"))
   "Doc"
   (("d" describe-foo-at-point "thing-at-pt")
    ("f" describe-function "function")
    ("v" describe-variable "variable")
    ("i" info-lookup-symbol "info lookup"))))

(use-package mixed-pitch
  :hook
  ;; If you want it in all text modes:
  (text-mode . mixed-pitch-mode)
  :config
  (setq mixed-pitch-set-height t))

(after! org
  (setq org-hide-emphasis-markers t
        org-ellipsis " ▾ "
        org-startup-with-inline-images t
        org-image-actual-width '(300))
  (after! org-capture
    (advice-add 'org-capture-select-template :override #'org-capture-select-template-prettier)
    (advice-add 'org-mks :override #'org-mks-pretty)
    (setq doct-after-conversion-functions '(+doct-iconify-capture-templates))
    (defun set-org-capture-templates ()
      (setq org-capture-templates
            (doct `(("Slipbox" :keys "s"
                     :icon ("inbox" :set "octicon" :color "green")
                     :file "digital-garden/org/inbox.org"
                     :type entry
                     :template ("* %?\n"))
                    ("Personal todo" :keys "t"
                     :icon ("checklist" :set "octicon" :color "green")
                     :file +org-capture-todo-file
                     :prepend t
                     :headline "Inbox"
                     :type entry
                     :template ("* TODO %?"
                                "%i %a"))
                    ("Personal note" :keys "n"
                     :icon ("sticky-note-o" :set "faicon" :color "green")
                     :file +org-capture-todo-file
                     :prepend t
                     :headline "Inbox"
                     :type entry
                     :template ("* %?"
                                "%i %a"))
                    ("Email" :keys "e"
                     :icon ("envelope" :set "faicon" :color "blue")
                     :file +org-capture-todo-file
                     :prepend t
                     :headline "Inbox"
                     :type entry
                     :template ("* TODO %^{type|reply to|contact} %\\3 %? :email:"
                                "Send an email %^{urgency|soon|ASAP|anon|at some point|eventually} to %^{recipient}"
                                "about %^{topic}"
                                "%U %i %a"))
                    ("Interesting" :keys "i"
                     :icon ("eye" :set "faicon" :color "lcyan")
                     :file +org-capture-todo-file
                     :prepend t
                     :headline "Interesting"
                     :type entry
                     :template ("* [ ] %{desc}%? :%{i-type}:"
                                "%i %a")
                     :children (("Webpage" :keys "w"
                                 :icon ("globe" :set "faicon" :color "green")
                                 :desc "%(org-cliplink-capture) "
                                 :i-type "read:web")
                                ("Article" :keys "a"
                                 :icon ("file-text" :set "octicon" :color "yellow")
                                 :desc ""
                                 :i-type "read:reaserch")
                                ("Information" :keys "i"
                                 :icon ("info-circle" :set "faicon" :color "blue")
                                 :desc ""
                                 :i-type "read:info")
                                ("Idea" :keys "I"
                                 :icon ("bubble_chart" :set "material" :color "silver")
                                 :desc ""
                                 :i-type "idea")))
                    ("Tasks" :keys "k"
                     :icon ("inbox" :set "octicon" :color "yellow")
                     :file +org-capture-todo-file
                     :prepend t
                     :headline "Tasks"
                     :type entry
                     :template ("* TODO %? %^G%{extra}"
                                "%i %a")
                     :children (("General Task" :keys "k"
                                 :icon ("inbox" :set "octicon" :color "yellow")
                                 :extra "")
                                ("Task with deadline" :keys "d"
                                 :icon ("timer" :set "material" :color "orange" :v-adjust -0.1)
                                 :extra "\nDEADLINE: %^{Deadline:}t")
                                ("Scheduled Task" :keys "s"
                                 :icon ("calendar" :set "octicon" :color "orange")
                                 :extra "\nSCHEDULED: %^{Start time:}t")))
                    ("Project" :keys "p"
                     :icon ("repo" :set "octicon" :color "silver")
                     :prepend t
                     :type entry
                     :headline "Inbox"
                     :template ("* %{time-or-todo} %?"
                                "%i"
                                "%a")
                     :file ""
                     :custom (:time-or-todo "")
                     :children (("Project-local todo" :keys "t"
                                 :icon ("checklist" :set "octicon" :color "green")
                                 :time-or-todo "TODO"
                                 :file +org-capture-project-todo-file)
                                ("Project-local note" :keys "n"
                                 :icon ("sticky-note" :set "faicon" :color "yellow")
                                 :time-or-todo "%U"
                                 :file +org-capture-project-notes-file)
                                ("Project-local changelog" :keys "c"
                                 :icon ("list" :set "faicon" :color "blue")
                                 :time-or-todo "%U"
                                 :heading "Unreleased"
                                 :file +org-capture-project-changelog-file)))
                    ("\tCentralised project templates"
                     :keys "o"
                     :type entry
                     :prepend t
                     :template ("* %{time-or-todo} %?"
                                "%i"
                                "%a")
                     :children (("Project todo"
                                 :keys "t"
                                 :prepend nil
                                 :time-or-todo "TODO"
                                 :heading "Tasks"
                                 :file +org-capture-central-project-todo-file)
                                ("Project note"
                                 :keys "n"
                                 :time-or-todo "%U"
                                 :heading "Notes"
                                 :file +org-capture-central-project-notes-file)
                                ("Project changelog"
                                 :keys "c"
                                 :time-or-todo "%U"
                                 :heading "Unreleased"
                                 :file +org-capture-central-project-changelog-file)))))))
    (set-org-capture-templates)
    (unless (display-graphic-p)
      (add-hook 'server-after-make-frame-hook
                (defun org-capture-reinitialise-hook ()
                  (when (display-graphic-p)
                    (set-org-capture-templates)
                    (remove-hook 'server-after-make-frame-hook
                                 #'org-capture-reinitialise-hook))))))
  (setq org-journal-dir "~/Documents/org/journal/"
        org-journal-date-prefix "* "
        org-journal-time-prefix "** "
        org-journal-date-format "%B %d, %Y (%A) "
        org-journal-file-format "%Y-%m-%d.org")
  (after! org-roam
    (add-hook 'org-roam-capture-new-node-hook #'jc/tag-new-node-as-draft)
    (cl-defmethod org-roam-node-type ((node org-roam-node))
      "Return the TYPE of NODE."
      (condition-case nil
          (file-name-nondirectory
           (directory-file-name
            (file-name-directory
             (file-relative-name (org-roam-node-file node) org-roam-directory))))
        (error "")))
    (setq
     org-roam-capture-templates
     '(("m" "main" plain
        "%?"
        :if-new (file+head "main/${slug}.org"
                           "#+title: ${title}\n")
        :immediate-finish t
        :unnarrowed t)
       ("r" "reference" plain "%?"
        :if-new
        (file+head "reference/${slug}.org" "#+title: ${title}\n")
        :immediate-finish t
        :unnarrowed t)
       ("a" "article" plain "%?"
        :if-new
        (file+head "articles/${slug}.org" "#+title: ${title}\n#+filetags: :article:\n")
        :immediate-finish t
        :unnarrowed t))
     org-roam-node-display-template
     (concat "${type:15} ${title:*} " (propertize "${tags:10}" 'face 'org-tag))
     org-roam-mode-sections
     '((org-roam-backlinks-section :unique t)
       org-roam-reflinks-section
       org-roam-unlinked-references-section))))

(use-package! org-appear
  :hook (org-mode . org-appear-mode)
  :config
  (setq org-appear-autoemphasis t
        org-appear-autosubmarkers t
        org-appear-autolinks nil)
  ;; for proper first-time setup, `org-appear--set-elements'
  ;; needs to be run after other hooks have acted.
  (run-at-time nil nil #'org-appear--set-elements))

(use-package! org-ol-tree
  :commands org-ol-tree
  :init
  (map! :map org-mode-map
        :after org
        :localleader
        :desc "Outline" "O" #'org-ol-tree))

(use-package! org-remark
  :after org
  :init
  (require 'org-remark-global-tracking)
  (org-remark-global-tracking-mode +1)
  (map! "C-c n m" #'org-remark-mark)
  :config
  (map! :map org-remark-mode-map
        "C-c n o" #'org-remark-open
        "C-c n ]" #'org-remark-view-next
        "C-c n [" #'org-remark-view-prev
        "C-c n R" #'org-remark-remove))

(use-package! org-roam-desktop
  :after org
  :init
  (map! "C-c n d" #'org-roam-desktop
        "C-c n a" #'org-roam-desktop-node-add)
  :config
  (setq org-roam-desktop-basename   "*OR-Desk--"
        org-roam-desktop-directory  "~/Documents/org/digital-garden/org"))

(use-package! org-transclusion
  :commands org-transclusion-mode
  :init
  (map! :after org :map org-mode-map
        "<f12>" #'org-transclusion-mode))

(use-package! zetteldesk
  :after org-roam
  :config
  (setq zetteldesk-hydra-prefix (kbd "C-c z"))
  (zetteldesk-mode)
  (require 'zetteldesk-kb)
  (advice-add 'org-remark-highlight-mark :after #'zetteldesk-remark-highlight-advice))
